subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecHandle.java

```diff
@@ -231,7 +232,7 @@ public ExecHandle start() {
             setState(ExecHandleState.STARTING);
 
             execHandleRunner = new ExecHandleRunner(this, streamsHandler, processLauncher, executorFactory);
-            executor.execute(execHandleRunner);
+            executor.execute(new BuildOperationIdentifierPreservingRunnable(execHandleRunner));
```

Comment:
This was a tough call to make. This problem can be fixed in several places with various pros/cons:
1) Using `BuildOperationIdentifierPreservingRunnable`. We wrap `Runnable` as required to transfer the build operation id to another thread. Usually, this happens through an `Executor`. This solution is easy to apply and can be reused across the code base.
2) Custom code as suggested with `StreamsForwarder`/`ExecHandleRunner`. This solution expose `BuildOperationIdentifierRegistry`, a singleton, that I'm still not convince we should be using throughout the code base.
3) Modify `StoppableExecutorImpl` to always preserve the build operation id. This most likely have a bigger impact on the performance as we don't always need to preserve the build operation.

I settle on option 1 as is give the best compromise between the other options. Given we discuss adding the possibility to have nested `BuildOperation` on workers, I assume the same may be true for threads inside the `Executor`. Passing the build operation id to the `exec` code during the creation of the object may yield incorrect operation id as the execution could be differed within another `BuildOperation`. The solution of this PR can also be used in [other scenario](https://github.com/gradle/gradle/pull/2052).